### ISSUE TITLE
python3.7 makes panic

### ISSUE THREAD

**Issue by rinrinne** [lambci/docker-lambda#193]
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.6 LTS
Release:	16.04
Codename:	xenial
$ docker --version
Docker version 18.09.6, build 481bc77
$ docker pull lambci/lambda:python3.7
$ docker inspect -f {{.Id}} lambci/lambda:python3.7
sha256:15ef123ad542b8f56614d34eabdca56c2568b4703ac7de526d22ff5ca95e5a57
$ docker inspect -f {{.Config.Entrypoint}} lambci/lambda:python3.7
[/var/rapid/init --bootstrap /var/runtime/bootstrap]
$ docker run -it --rm --entrypoint="" lambci/lambda:python3.7 /bin/bash
bash-4.2$ /var/rapid/init --bootstrap /var/runtime/bootstrap
START RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72 Version: $LATEST
END RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72
REPORT RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72	Duration: 0.00 ms	Billed Duration: 100 ms	Memory Size: 1536 MB	Max Memory Used: 9 MB	
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x6cca70]

goroutine 1 [running]:
main.(*ExitError).Error(0xc0001aa010, 0xc00019e0c0, 0xc0000b7b60)
	/app/init.go:540 +0x30
main.logEndRequest(0x7ed940, 0xc0001aa010)
	/app/init.go:394 +0x848
main.endInvoke(0x7ed940, 0xc0001aa010)
	/app/init.go:354 +0xce
main.abortRequest(...)
	/app/init.go:338
main.main()
	/app/init.go:224 +0x129d
bash-4.2$ 
```
Same result: run /var/rapid/init only.

**Comment by mhart** [comment]
What are you trying to do?

**Comment by mhart** [comment]
`lambci/lambda:python3.7` (or any of the other non-build containers) aren't intended to be run interactively

**Comment by rinrinne** [comment]
You can see the same result with the followings.
```
$ sam init --runtime python3.7
[+] Initializing project structure...

Project generated: ./sam-app

Steps you can take next within the project folder
===================================================
[*] Invoke Function: sam local invoke HelloWorldFunction --event event.json
[*] Start API Gateway locally: sam local start-api

Read sam-app/README.md for further instructions

[*] Project initialization is now complete
$ sam local invoke HelloWorldFunction --event event.json
...

```

**Comment by mhart** [comment]
```
$ sam local invoke HelloWorldFunction --event event.json
2019-06-13 10:36:01 Invoking app.lambda_handler (python3.7)

Fetching lambci/lambda:python3.7 Docker container image.......................................................................................................
2019-06-13 10:36:20 Mounting /Users/michael/github/docker-lambda/base/sam-app/hello_world as /var/task:ro,delegated inside runtime container
START RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72 Version: $LATEST
END RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72
REPORT RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72	Duration: 15.18 ms	Billed Duration: 100 ms	Memory Size: 128 MB	Max Memory Used: 22 MB	
{"statusCode":200,"body":"{\"message\": \"hello world\"}"}
```

**Comment by mhart** [comment]
What system are you running docker on?

**Comment by rinrinne** [comment]
Hmm...OK, Maybe environment issue?
```
$ uname -a
Linux minerva 4.15.0-50-generic #54~16.04.1-Ubuntu SMP Wed May 8 15:55:19 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.6 LTS
Release:	16.04
Codename:	xenial
$ docker --version
Docker version 18.09.6, build 481bc77
```

**Comment by rinrinne** [comment]
Docker image id in my environment is:
```
$ docker inspect -f {{.Id}} lambci/lambda:python3.7
sha256:15ef123ad542b8f56614d34eabdca56c2568b4703ac7de526d22ff5ca95e5a57
```

**Comment by mhart** [comment]
Hmmm. I don't really see anything obvious unfortunately.

Mine is `Server Version: 18.09.2` and `Kernel Version: 4.9.125-linuxkit` – and there's no `apparmor`

I haven't heard of this happening for anyone else (yet), so I think it must be something to do with your setup, but I can't think of what it would be.

Maybe try stopping `apparmor` and see if that makes a difference? Unlikely though

**Comment by rinrinne** [comment]
I removed system installed apparmor then try again, but unfortunately still same issue.

Maybe this is NPE issue in /var/rapid/init. Seems https://lambci.s3.amazonaws.com/fs/python3.7.tgz contains this.

**Comment by mhart** [comment]
That's actually overridden by this: https://github.com/lambci/docker-lambda/blob/master/provided/run/init.go

So the error is occurring around here: https://github.com/lambci/docker-lambda/blob/master/provided/run/init.go#L222-L226

The only thing I can think is that your docker setup isn't allowing the local http server to start. Are you restricting network calls in some way?

**Comment by rinrinne** [comment]
Thanks for your hint! Seem proxy setting within container.
I'm behind proxy, so need to proxy setting in container as followings (by ~/.docker/config.json)
```
http_proxy=http:/xxxxxxxx/
https_proxy=http:/xxxxxxxx/
no_proxy=localhost,127.0.0.1
```
After unset them all, /var/rapid/init to be normal. So I tested with the following combination.

|set variable|result|
|------------|-------|
| http_proxy | NG |
| https_proxy | OK |
|http_proxy + no_proxy   | NG |

Seems /var/rapid/init uses proxy without reflecting no_proxy.

**Comment by rinrinne** [comment]
I can solve this issue by workaround: remove proxy settings in ~/.docker/config.json before using sam.
But I will face related issue again if use AWS SDK within lambda due to no proxy setting...  (fortunately I have no plan to do that as of now)

Another workaround: use runtime python3.6.

**Comment by mhart** [comment]
I'll see if I can reproduce by setting those proxy settings

**Comment by mhart** [comment]
Can you try `lambci/lambda:python3.7-local` and see if that works with your `no_proxy` settings?

I changed `init` so that it only listens to `127.0.0.1`

**Comment by mhart** [comment]
If it works, I'll update all of the runtimes 👍

**Comment by rinrinne** [comment]
python3.7-local works fine! Great!

### Instructions
```
$ cat ~/.docker/config.json
{
  "proxies": {
    "default": {
      "httpProxy": "http://proxy.mydomain.net:8080",
      "httpsProxy": "http://proxy.mydomain.net:8080",
      "noProxy": "mydomain.net,localhost,127.0.0.1"
    }
  }
}
```
#### With image only
```
$ docker pull lambci/lambda:python3.7-local
$ docker run -it --rm --entrypoint="" lambci/lambda:python3.7-local /bin/bash
bash-4.2$ /var/rapid/init
```
Result: Expecting error (no panic!)

#### With SAM
```
$ docker tag lambci/lambda:python3.7 lambci/lambda:python3.7-orig
$ docker tag lambci/lambda:python3.7-local lambci/lambda:python3.7
$ sam init --runtime python3.7
$ cd sam-app
$ sam local invoke HelloWorldFunction --event event.json --skip-pull-image
2019-06-14 11:14:49 Invoking app.lambda_handler (python3.7)
2019-06-14 11:14:49 Found credentials in shared credentials file: ~/.aws/credentials
2019-06-14 11:14:49 Requested to skip pulling images ...

2019-06-14 11:14:49 Mounting /path/to/sam-app/hello_world as /var/task:ro,delegated inside runtime container
START RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72 Version: $LATEST
END RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72
REPORT RequestId: 52fdfc07-2182-154f-163f-5f0f9a621d72	Duration: 2.66 ms	Billed Duration: 100 ms	Memory Size: 128 MB	Max Memory Used: 23 MB	
{"statusCode":200,"body":"{\"message\": \"hello world\"}"}
```
Success !

**Comment by mhart** [comment]
Resolved in a549d675806ca7f24cdfc771b8c31f69da748299 – you should be able to use the normal `lambci/lambda:python3.7` image now 👍 

Thanks for reporting!
